console-subscriber/src/visitors.rs

```
@@ -0,0 +1,369 @@
+
+impl Visit for AsyncOpVisitor {
+    fn record_debug(&mut self, _: &field::Field, _: &dyn std::fmt::Debug) {}
+
+    fn record_str(&mut self, field: &tracing_core::Field, value: &str) {
+        if field.name() == Self::ASYNC_OP_SRC_FIELD_NAME {
+            self.source = Some(value.to_string());
+        }
+    }
+}
+
+impl WakerVisitor {
+    const WAKE: &'static str = "waker.wake";
+    const WAKE_BY_REF: &'static str = "waker.wake_by_ref";
+    const CLONE: &'static str = "waker.clone";
+    const DROP: &'static str = "waker.drop";
+    const TASK_ID_FIELD_NAME: &'static str = "task.id";
+
+    pub(crate) fn result(self) -> Option<(span::Id, WakeOp)> {
+        self.id.zip(self.op)
+    }
+}
+
+impl Visit for WakerVisitor {
+    fn record_debug(&mut self, _: &field::Field, _: &dyn std::fmt::Debug) {
+        // don't care (yet?)
+    }
+
+    fn record_u64(&mut self, field: &tracing_core::Field, value: u64) {
+        if field.name() == Self::TASK_ID_FIELD_NAME {
+            self.id = Some(span::Id::from_u64(value));
+        }
+    }
+
+    fn record_str(&mut self, field: &tracing_core::Field, value: &str) {
+        if field.name() == "op" {
+            self.op = Some(match value {
+                Self::WAKE => WakeOp::Wake,
+                Self::WAKE_BY_REF => WakeOp::WakeByRef,
+                Self::CLONE => WakeOp::Clone,
+                Self::DROP => WakeOp::Drop,
+                _ => return,
+            });
+        }
+    }
+}
+
+impl PollOpVisitor {
+    pub(crate) const POLL_OP_EVENT_NAME: &'static str = "runtime.resource.poll_op";
+    const OP_NAME_FIELD_NAME: &'static str = "op_name";
+    const OP_READINESS_FIELD_NAME: &'static str = "readiness";
+    const OP_READINESS_READY: &'static str = "ready";
+    const OP_READINESS_PENDING: &'static str = "pending";
+
+    pub(crate) fn result(self) -> Option<(String, Readiness)> {
+        let op_name = self.op_name?;
+        let readiness = self.readiness?;
+        Some((op_name, readiness))
```

I notice that this will silently fail if either `op_name` _or_ `readiness` is missing. Do we want to potentially handle incomplete data? What if a runtime decides to record a `poll_op` with a `name` but no `readiness` --- is that an error in the user code, or can we still do something useful with that?